### ISSUE TITLE
Bun & Deno Comparision

### ISSUE THREAD

**Issue by ronag** [nodejs/performance#55]
It would be nice to have some benchmarks vs Bun and Deno with things we would like to make faster. Does provide some incentive and makes it a bit more interesting to perf work and also a good indication of what things actually can be made faster

**Comment by benjamingr** [comment]
I agree, I also think it's important to measure actual user flows (kind of like what we did with boxbee for bluebird) and not micro-benchmarks.

**Comment by anonrig** [comment]
I've been using my own micro-benchmarks which helped me a lot: http://github.com/anonrig/node-benchmarks

**Comment by anonrig** [comment]
I can move my benchmark repository to this repository and start creating a suite of benchmarks against Bun & Deno and Node. wdyt? @ronag

**Comment by tniessen** [comment]
> I also think it's important to measure actual user flows (kind of like what we did with boxbee for bluebird) and not micro-benchmarks.

I agree with @benjamingr. It would be great to benchmark some realistic applications instead of only using micro-benchmarks, and then using profiling to find out why Node.js is slower for those particular applications (if it is). Micro-benchmarks are great for what they are, but they don't say anything about real-world impact of performance optimizations.

**Comment by nassau-t** [comment]
Just my 2 cents,

Regarding Bun, in their github https://github.com/oven-sh/bun, you can see
![image](https://github.com/nodejs/performance/assets/56589513/83ad0952-d196-4fa5-b6f2-9f77e234c890)

In the case of nodejs, in github https://github.com/nodejs/node, you can see
![image](https://github.com/nodejs/performance/assets/56589513/bb28c450-f6a9-4b0a-9917-303433f10fbc)

So, Bun has A LOT of things implemented in Zig ("similar" to C++).
 I don't know if this is the way to follow, but it's obvious that it's impossible to get same performance of Zig or C++ with javascript code.

**Comment by tniessen** [comment]
> it's obvious that it's impossible to get same performance of Zig or C++ with javascript code.

@nassau-t It's not as simple as you make it sound. With the V8 JavaScript engine, some operations are faster when implemented in JavaScript as opposed to C++, primarily because of significant overheads within the V8 engine when crossing the JavaScript/C++ boundary. This is why V8 added ["fast API calls"](https://github.com/nodejs/node/blob/cd97e28860d109e1576e05b4b79692aa013a4e6e/deps/v8/include/v8-fast-api-calls.h), which reduce said overhead, but due to technical constraints, they can only reasonably be used for a fraction of Node.js APIs so far.

Of course, it's no secret that Bun's performance advantage comes, at least in part, from its reliance on Zig and the tight integration with JavaScriptCore. This is not a low barrier to contributors, and more than 75 % of all commits so far have been authored by the same person.

Hypothetically, Node.js could move a lot of logic from JavaScript to C++, and we would likely see performance improvements in some areas. I am not generally opposed to this (and it would solve a variety of other issues), but it would raise the barrier to contributing to Node.js because Node.js tends to have more contributors who are comfortable working on JavaScript than contributors who like writing C++.

Anyway, this issue is about adding benchmarks for comparing the performance across JavaScript runtimes. It is not the right place to discuss why you think one runtime is faster than another. Please open a new issue if you'd like to share more thoughts.

**Comment by nassau-t** [comment]
> Anyway, this issue is about adding benchmarks for comparing the performance across JavaScript runtimes. It is not the right place to discuss why you think one runtime is faster than another. Please open a new issue if you'd like to share more thoughts.

Maybe doing benchmarks without thinking about what makes a thing faster than other it's a nonsense.

**Comment by tniessen** [comment]
> Maybe doing benchmarks without thinking about what makes a thing faster than other it's a nonsense.

How do you know which one is faster than the other without doing benchmarks in the first place?

**Comment by ronag** [comment]
This issue is about creating benchmarks. Please stay on topic.
